Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithmRZ.cpp

```diff
@@ -0,0 +1,193 @@
+/* Copyright 2019-2020 David Grote
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "PsatdAlgorithmRZ.H"
+#include "WarpXConst.H"
+
+#include <cmath>
+
+using amrex::operator""_rt;
+
+
+/* \brief Initialize coefficients for the update equation */
+PsatdAlgorithmRZ::PsatdAlgorithmRZ (SpectralHankelKSpace const & spectral_kspace,
+                                    amrex::DistributionMapping const & dm,
+                                    int const n_rz_azimuthal_modes, int const norder_z,
+                                    bool const nodal, amrex::Real const dt_step)
+     // Initialize members of base class
+     : SpectralBaseAlgorithmRZ(spectral_kspace, dm,
+                               norder_z, nodal),
+       dt(dt_step)
+{
+
+    // Allocate the arrays of coefficients
+    amrex::BoxArray const & ba = spectral_kspace.spectralspace_ba;
+    C_coef = SpectralCoefficients(ba, dm, n_rz_azimuthal_modes, 0);
+    S_ck_coef = SpectralCoefficients(ba, dm, n_rz_azimuthal_modes, 0);
+    X1_coef = SpectralCoefficients(ba, dm, n_rz_azimuthal_modes, 0);
+    X2_coef = SpectralCoefficients(ba, dm, n_rz_azimuthal_modes, 0);
+    X3_coef = SpectralCoefficients(ba, dm, n_rz_azimuthal_modes, 0);
+
+    coefficients_initialized = false;
+}
+
+/* Advance the E and B field in spectral space (stored in `f`)
+ * over one time step */
+void
+PsatdAlgorithmRZ::pushSpectralFields(SpectralFieldDataHankel & f)
+{
+
+    if (not coefficients_initialized) {
+        // This is called from here since it needs the kr values
+        // which can be obtained from the SpectralFieldDataHankel
```

Comment:
Maybe. Now there is a cross dependency since SpectralFieldDataRZ needs `algorithm->getRequiredNumberOfFields().` This could be removed since `algorithm->getRequiredNumberOfFields()` just returns `SpectralFieldIndex::n_fields` which is defined in SpectralFieldData.H anyway, so SpectralFieldDataRZ would have direct access to it. I think using `SpectralFieldIndex::n_fields` directly would be cleaner anyway. This follows the same pattern as in the Cartesian code.